Create a fork of go-yaml

natasha41575 opened the pull request:
This PR creates forks of go-yaml v2 and go-yaml v3.

Related issue: https://github.com/kubernetes-sigs/yaml/issues/72

This forks and also contains the changes in https://github.com/go-yaml/yaml/pull/753 that are necessary for kustomize (and are the reason we need to make the fork).

I've added @KnVerey and myself as owners; as discussed in meeting we are happy to maintain this. I also added the disclaimer stating that we will not accept large changes here.

natasha41575 commented:
/cc @KnVerey 
/cc @liggitt 
/cc @lavalamp

natasha41575 commented:
The linter is complaining about the code style of the go-yaml code, is there a way to ignore the forked directories?

liggitt commented on thirdparty/github.com/go-yaml/OWNERS:
```diff
@@ -0,0 +1,28 @@
+# See the OWNERS docs at https://go.k8s.io/owners
```
if these packages are not going to be referenced outside this library, they should be under an `internal` subpackage

if they are going to be referenced outside this module, I wouldn't expect callers to have to use import paths containing thirdparty and irrelevant github.com subpackage names

I was imagining something more like sigs.k8s.io/yaml/goyaml.v2 or something similar

natasha41575 replied:
> if these packages are not going to be referenced outside this library, they should be under an internal subpackage

We (kustomize) need them to be available outside this library. I will rename their import paths to `sigs.k8s.io/yaml/goyaml.v2` and `sigs.k8s.io/yaml/goyaml.v3`

liggitt commented on thirdparty/github.com/go-yaml/yaml.v2/LICENSE:
```diff
@@ -0,0 +1,201 @@
+                                 Apache License
```
It's good to include the license, but I don't think this bubbles up to the root and gets picked up by dependents like we need it to. We may need to include this in the root LICENSE file as well

natasha41575 replied:
Do I just need to append the contents of this file to the root LICENSE?

natasha41575 replied:
Based on conversation, I updated the root license by appending the forked yaml libs' licenses' content to the root license, hopefully that is enough for now.

liggitt commented:
It might be worth setting up apidiff presubmits on the go yaml subpackage trees to ensure we don't break go API compatibility on changes accidentally, since I expect these sub packages to attract a lot of usage. xref https://github.com/kubernetes/utils/blob/master/.github/workflows/test.yml#L52

liggitt commented on thirdparty/github.com/go-yaml/yaml.v3/yaml.go:
```diff
@@ -1,698 +1,808 @@
-// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
-// See the License for the specific language governing permissions and
-// limitations under the License.
+// Copyright (c) 2006-2010 Kirill Simonov
```
curious where this diff came from

natasha41575 replied:
May have been a copy-paste error; you're right that there shouldn't be a copyright diff in this commit. Will update

liggitt commented on thirdparty/github.com/go-yaml/yaml.v3/yamlh.go:
```diff
@@ -1,807 +1,707 @@
 //
 // Copyright (c) 2011-2019 Canonical Ltd
-// Copyright (c) 2006-2010 Kirill Simonov
```
same question here… I expected this commit to have minimal changes over the yaml.v3 copy

natasha41575 commented:
@liggitt I have addressed most of your comments, but still have a couple of questions:

1. How to add the licenses from go-yaml to the root license? Should I just append the contents, or is there a particular delimiter or heading I should use?

2. To set up the api-diff, would there be any issues with copying the script https://github.com/kubernetes/utils/blob/master/hack/verify-apidiff.sh into this repo?

liggitt commented:
> * How to add the licenses from go-yaml to the root license? Should I just append the contents, or is there a particular delimiter or heading I should use?

I'm... not sure... the file at goyaml.v3/LICENSE already enumerates multiple licenses and specific files. Maybe that's an example to follow, but I really don't have expertise in the implications/requirements here. @dims / @cblecker - any insight here? We want to make sure we're complying with licenses of code being included, and also minimize issues with license scanners, etc, from folks consuming the project.

> 2. To set up the api-diff, would there be any issues with copying the script https://github.com/kubernetes/utils/blob/master/hack/verify-apidiff.sh into this repo?

Yeah, copy whatever is useful. Will probably need some tweaks to make it pay attention just to the two go yaml subpackages

liggitt commented on goyaml.v3/encode_test.go:
```diff
@@ -19,6 +19,7 @@ import (
 	"bytes"
 	"fmt"
 	"math"
+	yaml "sigs.k8s.io/yaml/goyaml.v3"
```
nit: spurious import reordering? this doesn't seem like it belongs with stdlib imports

natasha41575 replied:
fixed

dims commented:
> I'm... not sure... the file at goyaml.v3/LICENSE already enumerates multiple licenses and specific files. Maybe that's an example to follow, but I really don't have expertise in the implications/requirements here. @dims / @cblecker - any insight here? We want to make sure we're complying with licenses of code being included, and also minimize issues with license scanners, etc, from folks consuming the project.

Let's start with what is already in goyaml.v3/LICENSE, if it trips up tools we use, we can then go figure out whether to fix the tool or fix the file.

cblecker commented:
We should consider using the third_party structure, similar to k/k: https://github.com/kubernetes/kubernetes/tree/master/third_party

Tooling is set up to handle that

liggitt commented:
that's where @natasha41575 started, but these are permanent forks that will be referenced by other downstreams, it's weird to expose a third_party package path to them - https://github.com/kubernetes-sigs/yaml/pull/76#discussion_r867390068

liggitt commented:
> That's part of the Apache 2.0 licence requirements to my understanding

Really? I've never see a directory structure requirement like that. I think pushing a "sigs.k8s.io/third_party/yaml.v2" import path to downstreams forever is *really* ugly, and sort of confusing, and would like to avoid it if at all possible. Not sure who could give a definitive answer on what's allowed here

dims commented:
@cblecker i haven't had that called out in AL 2.0, it's for our convenience we have the extra third_party so we can be careful when we touch code in there.

I agree with @liggitt that this is too ugly to impose on downstream folks.

@liggitt i've opened https://github.com/cncf/foundation/issues/354 until there is direction otherwise, we can keep doing what is best for folks that use us.

cblecker commented:
@dims sorry, I guess I skipped a mental step in my thought process

So Apache 2.0 requires, in section 4(b), that any modifications have prominent notices that they've been modified. So third_party, like vendor, calls out where code is copied into this repo so it can be treated differently (with additional notices on modification).

Again though, this is my "IANAL" understanding :)

dims commented:
@cblecker no worries, i got it on record from ChrisA in https://github.com/cncf/foundation/issues/354 we are good to go here.

liggitt commented:
also seems worth noting somewhere what version of yaml.v3 this is synced to, and pulling in https://github.com/go-yaml/yaml/releases/tag/v3.0.1 in this first sync

natasha41575 commented:
Thanks everyone for the feedback and looking into the license issue! I will update the PR based on the conversation here

natasha41575 commented on .github/workflows/go.yml:
```diff
@@ -22,7 +22,7 @@ jobs:
-      run: diff -u <(echo -n) <(golint $(go list -e ./...) | grep -v YAMLToJSON)
+      run: diff -u <(echo -n) <(golint $(go list -e ./... | grep -v /goyaml) | grep -v YAMLToJSON)
     - name: Run go vet
       run: GO111MODULE=on go vet .
```
Yeah there are a lot of existing vet errors on them but interestingly only in the test files

natasha41575 commented on goyaml.v2/decode_test.go:
```diff
@@ -697,7 +697,7 @@ var unmarshalTests = []struct {
 	}, {
 		"a: 123456E1\n",
-		M{"a": 123456E1},
+		M{"a": 123456e1},
```
done

natasha41575 commented:
/retest

natasha41575 commented:
> also seems worth noting somewhere what version of yaml.v3 this is synced to, and pulling in https://github.com/go-yaml/yaml/releases/tag/v3.0.1 in this first sync

 Done

natasha41575 commented:
@liggitt This PR is updated to address all your comments. The exception is that I have not set up an api-diff, but can do so next week if that's a blocker here.

natasha41575 commented:
@liggitt Do you have any idea what's causing the CI failures? For some reason something seems to be trying to download `gopkg.in/yaml.v2`... even though I've removed all of its references to point at the fork instead.

This doesn't happen locally, all tests, apidiff, and builds seem to be working fine on my machine. But CI has this error:

```
Run GO111MODULE=on go vet .
  GO111MODULE=on go vet .
  shell: /usr/bin/bash -e {0}
  env:
    GOROOT: /opt/hostedtoolcache/go/1.19.0/x64
Error: bench_test.go:[2](https://github.com/kubernetes-sigs/yaml/runs/7941717856?check_suite_focus=true#step:7:2)[4](https://github.com/kubernetes-sigs/yaml/runs/7941717856?check_suite_focus=true#step:7:4):2: no required module provides package gopkg.in/yaml.v2; to add it:
	go get gopkg.in/yaml.v2
Error: Process completed with exit code 1.
```

I can make the build job errors go away in CI by adding `gopkg.in/yaml.v2 //indirect` in the go.mod file but I don't know why it needs that (I feel like it shouldn't?). Even if I do that, I still get the apidiff error in CI that I can't reproduce locally.

natasha41575 commented:
/test Go / apidiff

natasha41575 commented:
@liggitt Can I set up the apidiff in a followup PR? I wrote one up, but then realized that if I add it to this PR, it would check the old api (which is nothing because there is no go-yaml fork in the old api) vs the api introduced in this PR, and fail accordingly.

Other than that, this PR is ready to go.

liggitt commented:
> @liggitt Can I set up the apidiff in a followup PR? I wrote one up, but then realized that if I add it to this PR, it would check the old api (which is nothing because there is no go-yaml fork in the old api) vs the api introduced in this PR, and fail accordingly.

yeah, a close follow up is fine

liggitt reviewed (commented):
structure lgtm, had a couple questions on the indent compatibility change commit (mostly trying to understand it)

liggitt commented on goyaml.v3/emitterc.go:
```diff
@@ -1144,8 +1149,11 @@ func yaml_emitter_process_head_comment(emitter *yaml_emitter_t) bool {
-func yaml_emitter_process_line_comment(emitter *yaml_emitter_t) bool {
+func yaml_emitter_process_line_comment(emitter *yaml_emitter_t, linebreak bool) bool {
 	if len(emitter.line_comment) == 0 {
+		if linebreak && !put_break(emitter) {
+			return false
+		}
```
what is this changing from yaml.v3 default behavior? is it configurable?

natasha41575 replied:
We have an issue that describes in detail what's being changed https://github.com/kubernetes-sigs/kustomize/issues/4033

We aren't changing anything about yaml.v3 default behavior, but we are adding an option to configure the style of sequence indents.

In older versions of yaml v3, we had:
```
foo:
- bar
```

Then yaml v3 made what kustomize considers a breaking change:

```
foo:
  - bar
```

We are leaving the default in tact but adding the option to configure this style so that we can pick up fixes upstream without introducing the breaking change to users. 

Kustomize is currently using its own internal fork to do this, and here is how we configure it in kustomize:

https://github.com/kubernetes-sigs/kustomize/blob/23de1499c27b1d040e90c8094051fe960661196a/kyaml/yaml/alias.go#L55-L59

liggitt replied:
This doesn't look like it's related to the sequence indent issue... I don't see this looking at the configurable `compact_sequence_indent` field, and the newly introduced parameter seems to be hard-coded either to true or false, not be configurable

This looks related to TestNewLinePreserved and doesn't look configurable, is that correct?

natasha41575 replied:
Oh, my bad! You're right. It came from this commit: https://github.com/go-yaml/yaml/pull/756

Here is the related issue: https://github.com/go-yaml/yaml/issues/755

We ran into this regression in kustomize and needed to fix it.

liggitt replied:
for the patches we're carrying for compatibility reasons with yaml.v2 behavior, can we do the following:
1. Keep each change in its own commit (so in this PR, one commit for the newline issue and commit one for the indenting issue)
2. Add good comments about the scenario being addressed and an explanation of the logic being modified (similar to the request for a comment at https://github.com/kubernetes-sigs/yaml/pull/76#discussion_r951887043). Maybe I'm alone in finding this code difficult to page into working memory, but explaining the mechanics of the tweaks we're making will help our future selves if we have to revisit our change.
3. Add a unit test to the root package demonstrating v2 and our patched v3 behave identically

natasha41575 replied:
> Maybe I'm alone in finding this code difficult to page into working memory

Absolutely not, this code is _extremely_ confusing and the only reason I hadn't responded to your above comments is because I need to refresh my memory how this code works. 

Thank you for these suggestions, I will update this PR accordingly.

natasha41575 replied:
> Keep each change in its own commit (so in this PR, one commit for the newline issue and commit one for the indenting issue)

Done

> Add good comments about the scenario being addressed and an explanation of the logic being modified (similar to the request for a comment at Create a fork of go-yaml #76 (comment)). Maybe I'm alone in finding this code difficult to page into working memory, but explaining the mechanics of the tweaks we're making will help our future selves if we have to revisit our change.

I added comments! Hopefully they make it more clear what is happening

> Add a unit test to the root package demonstrating v2 and our patched v3 behave identically

I'm not entirely sure what you're looking for here. `yaml.v2` and `yaml.v3` upstream behave differently (for example, yaml.v2 does the "compact" sequence indent by default while yaml.v3 does not). There are some other marshaling differences between the two as well. 

I added a unit test demonstrating that the upstream yaml.v3 and our patched fork of it behave identically.

liggitt replied:
> > Add a unit test to the root package demonstrating v2 and our patched v3 behave identically
> 
> I'm not entirely sure what you mean by this. `yaml.v2` and `yaml.v3` upstream behave differently (for example, yaml.v2 does the "compact" sequence indent by default while yaml.v3 does not). There are some other subtle marshaling and unmarshalling differences between the two as well.

As we're planning to switch this library's use to our fork of yaml.v3, we would want to make sure we could configure it in a way that preserved yaml.v2 behavior. For each difference we patched yaml.v3 to make consistent with yaml.v2, a test in the root package of this library to demonstrate that patch would be good. Something of this form:

1. setup
2. marshal or unmarshal with yaml.v2
3. marshal or unmarshal with yaml.v3 (maybe with non-default settings)
4. verify result is identical

liggitt replied:
> I added a unit test demonstrating that the upstream yaml.v3 and our patched fork of it behave identically, if that helps.

That wasn't really what I meant, and requires adding a dependency back on upstream yaml.v3, which I don't think we want. I'm more interested in demonstrating consistency between the existing yaml.v2 and our patched yaml.v3

natasha41575 replied:
Ah, that makes more sense! Apologies for the confusion. I have updated the test accordingly to do what you described above - it verifies that if you use the non-default settings on the yaml.v3 library that I added (compact sequence option, and setting the encoder indent), that yaml.v2 and yaml.v3 behave identically.

natasha41575 commented on goyaml.v3/emitterc.go:
```diff
@@ -243,6 +243,9 @@ func yaml_emitter_increase_indent(emitter *yaml_emitter_t, flow, indentless bool
 			emitter.indent = emitter.best_indent * ((emitter.indent + emitter.best_indent) / emitter.best_indent)
 		}
+		if compact_seq {
+			emitter.indent = emitter.indent - 2
```
Yeah, IIRC it needs to subtract 2 in both situations. 

The value `compact_seq` passed in is always set to false when this function is called, except when we are dealing with sequence nodes. So this gets triggered to subtract 2 only when we are increasing the indent to account for sequence nodes, which will be correct because we need to subtract 2 to account for the `- ` at the beginning of the sequence node. 

I will add this as a comment.

natasha41575 commented on goyaml.v3/emitterc.go:
```diff
@@ -728,7 +731,9 @@ func yaml_emitter_emit_flow_mapping_value(emitter *yaml_emitter_t, event *yaml_e
 func yaml_emitter_emit_block_sequence_item(emitter *yaml_emitter_t, event *yaml_event_t, first bool) bool {
 	if first {
-		if !yaml_emitter_increase_indent(emitter, false, false) {
+		seq := emitter.mapping_context && (emitter.column == 0 || !emitter.indention) &&
```
Added a comment

natasha41575 commented on yaml_test.go:
```diff
@@ -654,7 +657,7 @@ func TestYAMLToJSON(t *testing.T) {
 			json: `{"t":"this is very long line with spaces and it must be longer than 80 so we will repeat that it must be longer that 80"}`,
 		},
-		"empty yaml value": {
+		"empty yamlv2 value": {
```
Whoops, I'll fix these

reinvantveer commented:
@liggitt are you good to move forward on this issue? Are there still unresolved tasks/review items here?

reinvantveer commented:
I could offer my help but haven't written a single line of Go code yet, so I doubt I can be of any service here.

liggitt commented:
> @liggitt are you good to move forward on this issue? Are there still unresolved tasks/review items here?

Yeah, the two comments starting at https://github.com/kubernetes-sigs/yaml/pull/76#discussion_r971941039 about how we test the patches we carry without adding a dependency back onto upstream yaml.v3

horacimacias commented:
I'm willing to help here; if somebody points me to a specific task/test that needs updating I can give a hand.

natasha41575 commented:
@horacimacias You're welcome to create another PR building off my branch if you'd like. I may be able to find some time in the next couple weeks to update this one but I'm not certain about that

horacimacias commented:
thanks @natasha41575 . Can you please give me a bit of information as to what is missing/failing or where can I help? if there is anything specific (test/feature/fix) I should be able to be more helpful.

natasha41575 commented:
/reopen

natasha41575 commented:
Sending an update here since a few people have pinged me about it — this has been in my backlog for a little while but I’m planning to prioritize this in the coming weeks.

natasha41575 commented:
> Yeah, the two comments starting at #76 (comment) about how we test the patches we carry without adding a dependency back onto upstream yaml.v3

Addressed this and rebased. Thanks!

Sorry for taking such a long time to update this.

natasha41575 commented:
/remove-lifecycle rotten

liggitt commented:
> @liggitt Addressed this and rebased. Thanks!
> 
> Sorry for taking such a long time to update this.

no worries, should be able to wrap up review this week or next, thanks

natasha41575 commented:
kind ping @liggitt, in case this fell off your radar

100mik commented:
Hello 👋🏼 
Just ran into this. This is something we at Carvel might find helpful at some point as well! Thank you so much for spearheading this @natasha41575 .
We have been hesitant to bump to yaml.v3 in [carvel-dev/ytt](https://github.com/carvel-dev/ytt) because of the breaking changes. This could be very helpful 🙏🏼 

(Sorry for the noise 🙇🏼 )

natasha41575 commented:
@liggitt pulled in your suggested changes

liggitt commented on goyaml.v3/emitterc.go:
```diff
@@ -239,9 +239,23 @@ func yaml_emitter_increase_indent(emitter *yaml_emitter_t, flow, indentless bool
 		if emitter.states[len(emitter.states)-1] == yaml_EMIT_BLOCK_SEQUENCE_ITEM_STATE {
 			// The first indent inside a sequence will just skip the "- " indicator.
 			emitter.indent += 2
+			if compact_seq {
+				// The value compact_seq passed in is almost always set to `false` when this function is called,
+				// except when we are dealing with sequence nodes. So this gets triggered to subtract 2 only when we
+				// are increasing the indent to account for sequence nodes, which will be correct because we need to
+				// subtract 2 to account for the - at the beginning of the sequence node.
+				emitter.indent = emitter.indent - 2
+			}
```
I still can't figure out how to exercise this path in a unit test

<img src="https://user-images.githubusercontent.com/980082/235240526-6cbc23ac-1aad-4f02-a65e-949f0b1dcdf7.png">

if we can't exercise this path, let's drop it for now, rather than possibly modify indentation in a way that will break something. We can always add it (with a test) if we figure out whether it's needed in this path as well.

Adjust dropping this in the `make sequence style configurable` commit

natasha41575 replied:
Done.

liggitt commented:
testing this in k/k in https://github.com/kubernetes/kubernetes/pull/121406 (the dependencies linter is expected to fail while it's pointing at a temp branch, but the other tests should pass)

liggitt commented:
looks like the [unit test failures](https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/121406/pull-kubernetes-unit/1715444595564744704) in https://github.com/kubernetes/kubernetes/pull/121406 are due to messages changes already present at HEAD (before this PR) added in https://github.com/kubernetes-sigs/yaml/pull/65/files#diff-acfe93d173e77894feaede834d0927584abe0fc551dba248d191e16edaa63611

some of those are just wrong (like the doubled `error converting YAML to JSON: error converting YAML to JSON: yaml: line 13: could not find expected ':'` bit)

others are just changing messages without necessarily improving things:

```
expected err: 
        strict decoding error: yaml: unmarshal errors:
          line 7: key "num" already set in map
got err: 
        strict decoding error: error converting YAML to JSON: yaml: unmarshal errors:
          line 7: key "num" already set in map
```

I'd like to back out some of the extra verbosity, but I can't tell if kustomize has taken a hard dependency on those strings existing in errors (https://github.com/kubernetes-sigs/kustomize/blob/0f2618b21d3f08c2b2c4ecced292469747ad3354/api/internal/kusterr/yamlformaterror.go#L49-L51)

liggitt commented:
/lgtm
/approve

all relevant tests in https://github.com/kubernetes/kubernetes/pull/121406 passed on this

natasha41575 commented:
@rfay I think we will need to wait for a release.

Then you will have to replace all instances of `import "gopkg.in/yaml.v3"` to `import `"sigs.k8s.io/yaml/goyaml.v3"` with a release that includes this fork and it should work.

https://github.com/kubernetes-sigs/kustomize/pull/5412/commits/11b76be57aa9b4887de2bb7fbb0764def1221957 shows what this will end up looking like in kustomize.

reinvantveer commented:
@natasha41575 @liggitt and all involved: congrats and thank you for your efforts. I believe this is a vital step in the k8s ecosystem allowing it to move towards standardisation of the primary means to deploy k8s resources: through yaml manifests. Yaml that conforms to a yaml standard, instead of a goyaml "dialect". The path on how to maintain backwards compatibility isn't quite clear yet, but this is a good step.
